experimental/displaynames/src/displaynames.rs

```diff
@@ -399,90 +418,207 @@ impl LocaleDisplayNamesFormatter {
     );
 
     /// Returns the display name of a locale.
+    /// This implementation is based on the algorithm described in
+    /// https://www.unicode.org/reports/tr35/tr35-general.html#locale_display_name_algorithm
+    ///
     // TODO: Make this return a writeable instead of using alloc
     pub fn of<'a, 'b: 'a, 'c: 'a>(&'b self, locale: &'c Locale) -> Cow<'a, str> {
-        // https://www.unicode.org/reports/tr35/tr35-general.html#Display_Name_Elements
+        let longest_matching_subtag = find_longest_matching_subtag(&locale, &self);
 
-        // TODO: This binary search needs to return the longest matching found prefix
-        // instead of just perfect matches
-        if let Some(displayname) = match self.options.style {
-            Some(Style::Short) => self
-                .locale_data
-                .get()
-                .short_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Long) => self
-                .locale_data
-                .get()
-                .long_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            Some(Style::Menu) => self
-                .locale_data
-                .get()
-                .menu_names
-                .get_by(|bytes| locale.strict_cmp(bytes).reverse()),
-            _ => None,
+        // Step - 1: Construct a locale display name string (LDN).
+        // Find the displayname for the longest_matching_subtag which was derived above.
+        let ldn = get_locale_display_name(&locale, &longest_matching_subtag, &self);
+
+        // Step - 2: Construct a vector of longest qualifying substrings (LQS).
+        // Find the displayname for the remaining locale if exists.
+        let lqs = get_longest_qualifying_substrings(&locale, &longest_matching_subtag, &self);
+
+        // Step - 3: Return the displayname based on the size of LQS.
+        if lqs.len() == 0 {
+            return ldn.to_string().into();
+        } else {
+            return Cow::Owned(alloc::format!("{} ({})", ldn, lqs.join(", ")));
         }
```

Comment:
A couple things:
* `ldn.to_string().into()` allocates a `String` and wraps it into `Cow::Owned`. `Cow::Borrowed(ldn)` is what we want
* `lsq.join` allocates an intermediate `String`, which we want to avoid

```suggestion
        let mut result = Cow::Borrowed(ldn);
        if lqs.len() > 0 {
            let mut output = String::with_capacity(result.len() + " (".len() + lqs.iter().map(|s| ", ".len() + s.len()).sum() - ", ".len() + ")".len());
            output.push_str(result);
            output.push_str(" (");
            output.push_str(lqs[0]);
            for lqs in lqs[1..] {
                output.push_str(", ");
                output.push_str(lqs);
            }
            output.push_str(")");
            result = Cow::Owned(output);
        }
        result
```

Reply:
Thanks for the suggestion. Incorporated it with a few minor changes.